tools/testtool/main.go

```diff
@@ -0,0 +1,93 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+
+	"dagger.io/dagger"
+)
+
+func main() {
+	if len(os.Args) < 3 {
+		panic("too few args")
+	}
+
+	ctx := context.Background()
+	client, err := dagger.Connect(ctx, dagger.WithLogOutput(os.Stdout))
+	if err != nil {
+		panic(err)
+	}
+	defer client.Close()
+
+	var kernel *dagger.File
+	var kpath, qemuCmd string
+
+	switch guestGoarch() {
+	case "amd64":
+		kpath = "/bzImage"
+		qemuCmd = "qemu-system-x86_64 -L %s -m 1G"
+		kernel = client.Container().From("ghcr.io/hugelgupf/vmtest/kernel-amd64:main").File(kpath)
+	case "arm":
+		kpath = "/zImage"
+		qemuCmd = "qemu-system-aarch64 -machine virt -cpu max -m 1G -L %s"
+		kernel = client.Container().From("ghcr.io/hugelgupf/vmtest/kernel-arm:main").File(kpath)
+	case "arm64":
+		kpath = "/Image"
+		qemuCmd = "qemu-system-arm -M virt,highmem=off -L %s"
+		kernel = client.Container().From("ghcr.io/hugelgupf/vmtest/kernel-arm64:main").File(kpath)
+	}
+
+	tmp, err := os.MkdirTemp(".", "ci-testing")
+	if err != nil {
+		panic(err)
+	}
+	defer os.RemoveAll(tmp)
+
+	qemuCmd = fmt.Sprintf(qemuCmd, filepath.Join(tmp, "zqemu", "pc-bios"))
+
+	if ok, err := client.
+		Container().
+		From("ghcr.io/hugelgupf/vmtest/qemu:main").
+		WithFile(kpath, kernel).
+		Directory("/").
+		Export(ctx, tmp); !ok || err != nil {
+		panic(err)
+	}
+
+	pwd, err := os.Getwd()
+	if err != nil {
+		panic(err)
+	}
+
+	kpath, err = filepath.Abs(filepath.Join(tmp, kpath))
+	if err != nil {
+		panic(err)
+	}
+
+	cmd := exec.Command(os.Args[1], os.Args[2:]...)
+	cmd.Env = append(
+		os.Environ(),
+		fmt.Sprintf("UROOT_KERNEL=%s", kpath),
+		fmt.Sprintf("UROOT_QEMU=%s", qemuCmd),
+		fmt.Sprintf("UROOT_SOURCE=%s", pwd),
+	)
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	if err := cmd.Run(); err != nil {
+		panic(err)
+	}
+}
+
+func guestGoarch() string {
```

Comment:
I think we've been on this path before, I think a simple function called 
EnvWithDefault(env, dfault string) string
would be fine. 
e.g. 
EnvWithDefault("UROOT_TESTARCH", runtime.GOARCH)
would be more useful. It's this same function you have here, but less specialized, and it's actually clearer in the one call what you are really doing. You don't have to look at the function to see what it does.

so
switch guestGoarch() {
becomes
switch EnvWIthDefault("UROOT_TESTARCH", runtime.GOARCH)

which I personally find clearer.